File: lib/features/clan/models/clan_capital_history.dart
class CapitalHistoryItems {
  final List<CapitalHistoryItem> items;
  final String? clanTag;
  final CapitalStats? stats;

  CapitalHistoryItems({required this.items, required this.clanTag, this.stats});

  factory CapitalHistoryItems.fromJson(
      Map<String, dynamic> json, String clanTag, {Map<String, dynamic>? statsData}) {
    try {
      return CapitalHistoryItems(
        items: json['history'] != null 
            ? List<CapitalHistoryItem>.from(
                json['history'].map((x) => CapitalHistoryItem.fromJson(x)))
            : [],
        clanTag: clanTag,
        stats: statsData != null ? CapitalStats.fromJson(statsData) : null,
      );
    } catch (e) {
      print("Error parsing CapitalHistoryItems: $e");
      return CapitalHistoryItems.empty();
    }
  }

  factory CapitalHistoryItems.empty() {
    return CapitalHistoryItems(items: [], clanTag: "");
  }
}

class CapitalHistoryItem {
  String state;
  DateTime startTime;
  DateTime endTime;
  int capitalTotalLoot;
  int raidsCompleted;
  int totalAttacks;
  int enemyDistrictsDestroyed;
  int offensiveReward;
  int defensiveReward;
  List<RaidMember>? members;
  List<RaidAttackLog>? attackLog;

  CapitalHistoryItem({
    required this.state,
    required this.startTime,
    required this.endTime,
    required this.capitalTotalLoot,
    required this.raidsCompleted,
    required this.totalAttacks,
    required this.enemyDistrictsDestroyed,
    required this.offensiveReward,
    required this.defensiveReward,
    required this.members,
    required this.attackLog,
  });

  factory CapitalHistoryItem.fromJson(Map<String, dynamic> json) {
    return CapitalHistoryItem(
      state: json['state'],
      startTime: DateTime.parse(json['startTime']),
      endTime: DateTime.parse(json['endTime']),
      capitalTotalLoot: json['capitalTotalLoot'],
      raidsCompleted: json['raidsCompleted'],
      totalAttacks: json['totalAttacks'],
      enemyDistrictsDestroyed: json['enemyDistrictsDestroyed'],
      offensiveReward: json['offensiveReward'],
      defensiveReward: json['defensiveReward'],
      members: json['members'] != null
          ? List<RaidMember>.from(
              json['members'].map((x) => RaidMember.fromJson(x)))
          : [],
      attackLog: json['attackLog'] != null
          ? List<RaidAttackLog>.from(
              json['attackLog'].map((x) => RaidAttackLog.fromJson(x)))
          : [],
    );
  }
}

class RaidMember {
  String tag;
  String name;
  int attacks;
  int attackLimit;
  int bonusAttackLimit;
  int capitalResourcesLooted;

  RaidMember({
    required this.tag,
    required this.name,
    required this.attacks,
    required this.attackLimit,
    required this.bonusAttackLimit,
    required this.capitalResourcesLooted,
  });

  factory RaidMember.fromJson(Map<String, dynamic> json) {
    return RaidMember(
      tag: json['tag'],
      name: json['name'],
      attacks: json['attacks'],
      attackLimit: json['attackLimit'],
      bonusAttackLimit: json['bonusAttackLimit'],
      capitalResourcesLooted: json['capitalResourcesLooted'],
    );
  }
}

class RaidAttackLog {
  RaidDefender defender;
  int attackCount;
  int districtCount;
  int districtsDestroyed;
  List<District> districts;

  RaidAttackLog({
    required this.defender,
    required this.attackCount,
    required this.districtCount,
    required this.districtsDestroyed,
    required this.districts,
  });

  factory RaidAttackLog.fromJson(Map<String, dynamic> json) {
    return RaidAttackLog(
      defender: RaidDefender.fromJson(json['defender']),
      attackCount: json['attackCount'],
      districtCount: json['districtCount'],
      districtsDestroyed: json['districtsDestroyed'],
      districts: List<District>.from(
          json['districts'].map((x) => District.fromJson(x))),
    );
  }
}

class RaidDefender {
  String tag;
  String name;
  int level;
  Map<String, String> badgeUrls;

  RaidDefender({
    required this.tag,
    required this.name,
    required this.level,
    required this.badgeUrls,
  });

  factory RaidDefender.fromJson(Map<String, dynamic> json) {
    return RaidDefender(
      tag: json['tag'],
      name: json['name'],
      level: json['level'],
      badgeUrls: Map<String, String>.from(json['badgeUrls']),
    );
  }
}

class District {
  int id;
  String name;
  int districtHallLevel;
  int destructionPercent;
  int stars;
  int attackCount;
  int totalLooted;
  List<Attack>? attacks;

  District({
    required this.id,
    required this.name,
    required this.districtHallLevel,
    required this.destructionPercent,
    required this.stars,
    required this.attackCount,
    required this.totalLooted,
    required this.attacks,
  });

  factory District.fromJson(Map<String, dynamic> json) {
    return District(
      id: json['id'],
      name: json['name'],
      districtHallLevel: json['districtHallLevel'],
      destructionPercent: json['destructionPercent'],
      stars: json['stars'],
      attackCount: json['attackCount'],
      totalLooted: json['totalLooted'],
      attacks: json['attacks'] != null
          ? List<Attack>.from(json['attacks'].map((x) => Attack.fromJson(x)))
          : [],
    );
  }
}

class Attack {
  String tag;
  String name;
  int destructionPercent;
  int stars;

  Attack({
    required this.tag,
    required this.name,
    required this.destructionPercent,
    required this.stars,
  });

  factory Attack.fromJson(Map<String, dynamic> json) {
    return Attack(
      tag: json['attacker']['tag'],
      name: json['attacker']['name'],
      destructionPercent: json['destructionPercent'],
      stars: json['stars'],
    );
  }
}

class CapitalStats {
  final int totalLoot;
  final int totalAttacks;
  final int numberOfWeeks;
  final int totalRaids;
  final int totalDistrictsDestroyed;
  final int totalOffensiveRewards;
  final int totalDefensiveRewards;
  final double avgLootPerAttack;
  final double avgLootPerWeek;
  final double avgAttacksPerWeek;
  final double avgAttacksPerRaid;
  final double avgAttacksPerDistrict;
  final double avgOffensiveRewards;
  final double avgDefensiveRewards;
  final CapitalRaidSummary? bestRaid;
  final CapitalRaidSummary? worstRaid;

  CapitalStats({
    required this.totalLoot,
    required this.totalAttacks,
    required this.numberOfWeeks,
    required this.totalRaids,
    required this.totalDistrictsDestroyed,
    required this.totalOffensiveRewards,
    required this.totalDefensiveRewards,
    required this.avgLootPerAttack,
    required this.avgLootPerWeek,
    required this.avgAttacksPerWeek,
    required this.avgAttacksPerRaid,
    required this.avgAttacksPerDistrict,
    required this.avgOffensiveRewards,
    required this.avgDefensiveRewards,
    this.bestRaid,
    this.worstRaid,
  });

  factory CapitalStats.fromJson(Map<String, dynamic> json) {
    return CapitalStats(
      totalLoot: json['totalLoot'] ?? 0,
      totalAttacks: json['totalAttacks'] ?? 0,
      numberOfWeeks: json['numberOfWeeks'] ?? 0,
      totalRaids: json['totalRaids'] ?? 0,
      totalDistrictsDestroyed: json['totalDistrictsDestroyed'] ?? 0,
      totalOffensiveRewards: json['totalOffensiveRewards'] ?? 0,
      totalDefensiveRewards: json['totalDefensiveRewards'] ?? 0,
      avgLootPerAttack: (json['avgLootPerAttack'] as num?)?.toDouble() ?? 0.0,
      avgLootPerWeek: (json['avgLootPerWeek'] as num?)?.toDouble() ?? 0.0,
      avgAttacksPerWeek: (json['avgAttacksPerWeek'] as num?)?.toDouble() ?? 0.0,
      avgAttacksPerRaid: (json['avgAttacksPerRaid'] as num?)?.toDouble() ?? 0.0,
      avgAttacksPerDistrict: (json['avgAttacksPerDistrict'] as num?)?.toDouble() ?? 0.0,
      avgOffensiveRewards: (json['avgOffensiveRewards'] as num?)?.toDouble() ?? 0.0,
      avgDefensiveRewards: (json['avgDefensiveRewards'] as num?)?.toDouble() ?? 0.0,
      bestRaid: json['bestRaid'] != null ? CapitalRaidSummary.fromJson(json['bestRaid']) : null,
      worstRaid: json['worstRaid'] != null ? CapitalRaidSummary.fromJson(json['worstRaid']) : null,
    );
  }
}

class CapitalRaidSummary {
  final String startTime;
  final int capitalTotalLoot;
  final int totalRewards;
  final int raidsCompleted;
  final int totalAttacks;
  final int enemyDistrictsDestroyed;
  final double avgAttacksPerRaid;
  final double avgAttacksPerDistrict;

  CapitalRaidSummary({
    required this.startTime,
    required this.capitalTotalLoot,
    required this.totalRewards,
    required this.raidsCompleted,
    required this.totalAttacks,
    required this.enemyDistrictsDestroyed,
    required this.avgAttacksPerRaid,
    required this.avgAttacksPerDistrict,
  });

  factory CapitalRaidSummary.fromJson(Map<String, dynamic> json) {
    return CapitalRaidSummary(
      startTime: json['startTime'] ?? '',
      capitalTotalLoot: json['capitalTotalLoot'] ?? 0,
      totalRewards: json['totalRewards'] ?? 0,
      raidsCompleted: json['raidsCompleted'] ?? 0,
      totalAttacks: json['totalAttacks'] ?? 0,
      enemyDistrictsDestroyed: json['enemyDistrictsDestroyed'] ?? 0,
      avgAttacksPerRaid: (json['avgAttacksPerRaid'] as num?)?.toDouble() ?? 0.0,
      avgAttacksPerDistrict: (json['avgAttacksPerDistrict'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

File: expo/src/features/clan/models/clan-capital-history.ts
import { apiDateOrEpoch, int, number, record, records, string, type JsonRecord } from './parsing';

export class Attack {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly destructionPercent: number,
    readonly stars: number,
  ) {}
  static fromJson(json: JsonRecord) {
    const attacker = record(json.attacker);
    return new Attack(
      string(attacker.tag),
      string(attacker.name),
      int(json.destructionPercent),
      int(json.stars),
    );
  }
}

export class District {
  constructor(
    readonly id: number,
    readonly name: string,
    readonly districtHallLevel: number,
    readonly destructionPercent: number,
    readonly stars: number,
    readonly attackCount: number,
    readonly totalLooted: number,
    readonly attacks: readonly Attack[],
  ) {}
  static fromJson(json: JsonRecord) {
    return new District(
      int(json.id),
      string(json.name),
      int(json.districtHallLevel),
      int(json.destructionPercent),
      int(json.stars),
      int(json.attackCount),
      int(json.totalLooted),
      records(json.attacks).map(Attack.fromJson),
    );
  }
}

export class RaidDefender {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly level: number,
    readonly badgeUrls: Readonly<Record<string, string>>,
  ) {}
  static fromJson(json: JsonRecord) {
    const badges = Object.fromEntries(
      Object.entries(record(json.badgeUrls)).map(([key, value]) => [key, string(value)]),
    );
    return new RaidDefender(string(json.tag), string(json.name), int(json.level), badges);
  }
}

export class RaidAttackLog {
  constructor(
    readonly defender: RaidDefender,
    readonly attackCount: number,
    readonly districtCount: number,
    readonly districtsDestroyed: number,
    readonly districts: readonly District[],
  ) {}
  static fromJson(json: JsonRecord) {
    return new RaidAttackLog(
      RaidDefender.fromJson(record(json.defender ?? json.attacker)),
      int(json.attackCount),
      int(json.districtCount),
      int(json.districtsDestroyed),
      records(json.districts).map(District.fromJson),
    );
  }
}

export class RaidMember {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly attacks: number,
    readonly attackLimit: number,
    readonly bonusAttackLimit: number,
    readonly capitalResourcesLooted: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new RaidMember(
      string(json.tag),
      string(json.name),
      int(json.attacks),
      int(json.attackLimit),
      int(json.bonusAttackLimit),
      int(json.capitalResourcesLooted),
    );
  }
}

export class CapitalHistoryItem {
  constructor(
    readonly state: string,
    readonly startTime: Date,
    readonly endTime: Date,
    readonly capitalTotalLoot: number,
    readonly raidsCompleted: number,
    readonly totalAttacks: number,
    readonly enemyDistrictsDestroyed: number,
    readonly offensiveReward: number,
    readonly defensiveReward: number,
    readonly members: readonly RaidMember[],
    readonly attackLog: readonly RaidAttackLog[],
    readonly defenseLog: readonly RaidAttackLog[],
  ) {}
  static fromJson(json: JsonRecord) {
    return new CapitalHistoryItem(
      string(json.state),
      apiDateOrEpoch(json.startTime),
      apiDateOrEpoch(json.endTime),
      int(json.capitalTotalLoot),
      int(json.raidsCompleted),
      int(json.totalAttacks),
      int(json.enemyDistrictsDestroyed),
      int(json.offensiveReward),
      int(json.defensiveReward),
      records(json.members).map(RaidMember.fromJson),
      records(json.attackLog).map(RaidAttackLog.fromJson),
      records(json.defenseLog).map(RaidAttackLog.fromJson),
    );
  }
}

export class CapitalRaidSummary {
  constructor(
    readonly startTime: string,
    readonly capitalTotalLoot: number,
    readonly totalRewards: number,
    readonly raidsCompleted: number,
    readonly totalAttacks: number,
    readonly enemyDistrictsDestroyed: number,
    readonly avgAttacksPerRaid: number,
    readonly avgAttacksPerDistrict: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new CapitalRaidSummary(
      string(json.startTime),
      int(json.capitalTotalLoot),
      int(json.totalRewards),
      int(json.raidsCompleted),
      int(json.totalAttacks),
      int(json.enemyDistrictsDestroyed),
      number(json.avgAttacksPerRaid),
      number(json.avgAttacksPerDistrict),
    );
  }
}

export class CapitalStats {
  constructor(
    readonly totalLoot: number,
    readonly totalAttacks: number,
    readonly numberOfWeeks: number,
    readonly totalRaids: number,
    readonly totalDistrictsDestroyed: number,
    readonly totalOffensiveRewards: number,
    readonly totalDefensiveRewards: number,
    readonly avgLootPerAttack: number,
    readonly avgLootPerWeek: number,
    readonly avgAttacksPerWeek: number,
    readonly avgAttacksPerRaid: number,
    readonly avgAttacksPerDistrict: number,
    readonly avgOffensiveRewards: number,
    readonly avgDefensiveRewards: number,
    readonly bestRaid: CapitalRaidSummary | null,
    readonly worstRaid: CapitalRaidSummary | null,
  ) {}
  static fromJson(json: JsonRecord) {
    const best = record(json.bestRaid);
    const worst = record(json.worstRaid);
    return new CapitalStats(
      int(json.totalLoot),
      int(json.totalAttacks),
      int(json.numberOfWeeks),
      int(json.totalRaids),
      int(json.totalDistrictsDestroyed),
      int(json.totalOffensiveRewards),
      int(json.totalDefensiveRewards),
      number(json.avgLootPerAttack),
      number(json.avgLootPerWeek),
      number(json.avgAttacksPerWeek),
      number(json.avgAttacksPerRaid),
      number(json.avgAttacksPerDistrict),
      number(json.avgOffensiveRewards),
      number(json.avgDefensiveRewards),
      Object.keys(best).length ? CapitalRaidSummary.fromJson(best) : null,
      Object.keys(worst).length ? CapitalRaidSummary.fromJson(worst) : null,
    );
  }
}

export class CapitalHistoryItems {
  constructor(
    readonly items: readonly CapitalHistoryItem[],
    readonly clanTag: string | null,
    readonly stats: CapitalStats | null = null,
  ) {}

  static fromJson(json: JsonRecord, clanTag: string, statsData?: JsonRecord): CapitalHistoryItems {
    try {
      return new CapitalHistoryItems(
        records(json.history).map(CapitalHistoryItem.fromJson),
        clanTag,
        statsData ? CapitalStats.fromJson(statsData) : null,
      );
    } catch {
      return CapitalHistoryItems.empty();
    }
  }

  static empty(): CapitalHistoryItems {
    return new CapitalHistoryItems([], '');
  }
}
